src/libexpr/primops.cc

```diff
@@ -892,6 +892,15 @@ static RegisterPrimOp primop_trace({
 });
 
 
+/* Takes two arguments and evaluates to the second one. Used as the
+ * builtins.traceVerbose implementation when --trace-verbose is not enabled
+ */
+static void prim_second(EvalState & state, const Pos & pos, Value * * args, Value & v)
```

Comment:
I find it a bit annoying that this doesn’t evaluate its first argument, as it means that the same Nix expression can run correctly without `--trace-verbose` but fail with it (`builtins.traceVerbose (rec { x = x; }.x) "foo"` for example).

Could we instead have something that does a `coerceToString` on its first argument (and just discards it) to be sure that both evaluation modes have the exact same semantics?

(that would obviously mean that `traceVerbose` is never free, so there’s a tradeoff to make here)

Reply:
From one point of view - yes you are right. But in general, if something is unused - it should not be evaluated but I can definitely see it being misleading.

I am down for both solutions, would anyone else like to share their thoughts on this?

Reply:
I already commented on this in https://github.com/NixOS/nix/pull/4914#discussion_r759708467

Reply:
You could use `prim_seq` here, which does evaluate its first argument. But I think being lazy is better. Otherwise you have to worry about the runtime cost of a debug message (e.g. `builtins.traceVerbose "${big expensive expression}" ...`) even when they're disabled.

Reply:
I guess we are sticking with it being lazy. Resolving this conversation